crates/ecs/src/stream.rs

```diff
@@ -204,38 +190,37 @@ pub struct FrozenWorldDiff {
 }
 impl FrozenWorldDiff {
     pub fn merge(diffs: &[Self]) -> WorldDiffView<'_> {
-        // only keep the last of WorldChange::Set for given EntityId and component path
-        let mut overwritten: HashSet<(EntityId, &str)> = HashSet::new();
-        let mut rev_changes = Vec::new();
+        // indexes of the last SetComponents for each entity
+        let mut set_idx: HashMap<EntityId, usize> = HashMap::new();
+        // merged changes
+        let mut changes = Vec::new();
 
-        // going backwards because it's easier to keep the first instead of removing/overwriting the previous ones
-        for change in diffs
-            .iter()
-            .rev()
-            .flat_map(|diff| diff.changes.iter().rev())
-        {
-            if let WorldChange::Set(entity_id, entry) = change {
-                if let Some(path) = entry.vtable.path {
-                    if !overwritten.insert((*entity_id, path)) {
-                        // not inserted -> we already have a Set for this entity_id and path
-                        continue;
+        for change in diffs.iter().flat_map(|diff| diff.changes.iter()) {
+            if let WorldChange::SetComponents(entity_id, entity) = change {
+                if let Some(idx) = set_idx.get(entity_id) {
+                    if let Some(WorldChange::SetComponents(_, existing_entity)) =
+                        changes.get_mut(*idx).map(Cow::to_mut)
+                    {
+                        existing_entity.merge(entity.clone());
+                    } else {
+                        panic!("Expected SetComponents");
```

Comment:
Can you extend this panic/add a comment about the invariants that would prevent this panic from firing? (Or if it *can* fire, can you add more information about what to look out for?)

Reply:
It should actually be `unreachable!()`. I'll change it and add a comment.

Reply:
https://github.com/AmbientRun/Ambient/commit/e3318819325dd26ce1f317b39d4e72723b9ca10e